### PR TITLE
Improve `type_for_source` function behaviour

### PR DESCRIPTION
**Pull request by GlassOfWhiskey**
1. When a `source` or `outputSource` field contains a list of sources, return an `ArraySchema` object as the type for source.
2. Handle `linkMerge` and `pickValue` type modifiers directly in the `type_for_source` function.

### CODE REVIEW

**Review by mr-c — CHANGES_REQUESTED** [review]
What if there is a `linkMerge`? Can you add a test for that scenario and compare to what `cwltool` computes?

And for Workflow outputs, it shouldn't matter if there is are multiple entries in `outputSource`; the type is the declared type, nothings else.

**Comment by GlassOfWhiskey** [comment]
> What if there is a `linkMerge`? Can you add a test for that scenario and compare to what `cwltool` computes?

Here is exactly why I was trying to split type check logic in multiple small PRs. There are MANY cases in which returning the exact type for a source is a pain:
 - `linkMerge` (nested vs flattened)
 - `pickValue` (one and first vs all)
 - `scatter` in the source (nested vs flattened)
- `scatter` in the sink (nested vs flattened)

I was trying to construct everything one piece at a time, otherwise I'll get lost XD

> And for Workflow outputs, it shouldn't matter if there is are multiple entries in outputSource; the type is the declared type, nothings else.

Here I did not undestand what you mean. If I have

```yaml
class: Workflow
...
outputs:
  out:
    type: string
    outputSource:
      - step1/out
      - step2/out
```
the type of the source cannot be `string`. The checker should then throw a type error, and to do that it should call `type_for_source`, which returns an `ArraySchema`, and see that it is different from the declared `string` type.
Am I missing something?

File: tests/test_parser_utils.py [hunk 1]
```diff
@@ -442,3 +466,15 @@ def test_v1_2_type_for_source_with_id() -> None:
...
+
+def test_v1_2_type_for_source_with_multiple_entries() -> None:
+    """Test that the type is correctly inferred from a list of source ids with CWL v1.2."""
+    uri = Path(HERE / "../testdata/count-lines7-wf_v12.cwl").resolve().as_uri()
+    cwl_obj = load_document_by_uri(uri)
+    source_type = cwl_utils.parser.cwl_v1_2_utils.type_for_source(
+        cwl_obj, cwl_obj.steps[0].in_[0].source
+    )
+    assert isinstance(source_type, cwl_utils.parser.cwl_v1_2.ArraySchema)
+    assert source_type.items[0].type == "array"
```
Comment (GlassOfWhiskey) [inline tests/test_parser_utils.py:479]: If here we assert `len(source_type.items) == 1`, we are indirectly testing that comparison between CWL type objects (introduced [here](https://github.com/common-workflow-language/schema_salad/pull/593)) is working correctly. 

However, maybe it is better to open a different PR for that, to avoid superimposing things.

What do you think?

File: cwl_utils/cwl_v1_2_expression_refactor.py [hunk 2]
```diff
@@ -1900,36 +1904,17 @@ def traverse_step(
...
-                        input_source_id = inp.source.split("#")[-1]
-                        source_type = utils.param_for_source_id(
-                            step.run, input_source_id, parent
-                        )
+                    input_source_id = (
+                        [source.split("#")[-1] for source in inp.source]
+                        if isinstance(inp.source, MutableSequence)
+                        else inp.source.split("#")[-1]
+                    )
+                    source_type = utils.type_for_source(step.run, inp.source, parent)
```
Comment (mr-c) [inline cwl_utils/cwl_v1_2_expression_refactor.py:1917]: How is this working? `source_type` must be a `cwl.WorkflowInputParameter` but that is not possible from the proposed `utils.type_for_source`.

**Review by mr-c — CHANGES_REQUESTED** [review]
I see a lot of lines of code without test coverage, can you fix that?

**Comment by GlassOfWhiskey** [comment]
> I see a lot of lines of code without test coverage, can you fix that?

They are mainly corner cases. I identified a dead end in the `scatter` logic and removed it

**Comment by GlassOfWhiskey** [comment]
Some of them are corner cases. For example here

```yaml
if isinstance(source_type_items, cwl.ArraySchema):
    if isinstance(source_type_items.items, list):
        if "null" not in source_type_items.items:
            source_type_items.items.append("null")
    elif source_type_items.items != "null":
        source_type_items.items = ["null", source_type_items.items]
```

it checks an object of type
```yaml
type: array
items: "null"
```

that is probably purely theoretical

**Comment by GlassOfWhiskey** [comment]
If you want to increase coverage, probably the best way will be to run the type checker to validate all the CWL Conformance tests.
Don't know if it is possible/easy

**Comment by GlassOfWhiskey** [comment]
But wait because something has gone wrong with the rebase.
Let me check again

File: cwl_utils/parser/cwl_v1_2_utils.py [hunk 3]
```diff
@@ -68,28 +167,79 @@ def convert_stdstreams_to_files(clt: cwl.CommandLineTool) -> None:
...
+                        cur_type = cwl.ArraySchema(items=cur_type, type='array')
+                else:
+                    cur_type = cwl.ArraySchema(items=cur_type, type='array')
+            new_type.append(cur_type)
+    if len(new_type) == 1:
+        new_type = new_type[0]
+    if linkMerge == 'merge_nested':
+        return cwl.ArraySchema(items=new_type, type='array')
+    elif linkMerge == 'merge_flattened':
+        return merge_flatten_type(new_type)
```
[the remark this answers is not in the document]
Comment (GlassOfWhiskey) [reply cwl_utils/parser/cwl_v1_2_utils.py]: Yes you're right.
I just changed the code

**Review by mr-c — APPROVED** [review]
